### PR TITLE
BUG, ENH: Aggregate data set before performing RM-Anova if needed

### PR DESCRIPTION
**Pull request by hoechenberger**
If the data set contains more than one observation per subject and experimental condition, the current AnovaRM implementation will generate incorrect results.

This commit adds an extra aggregation step before performing the Anova and displays a warning. By default, the `np.mean` function will be used; but the user is free to specify a different function.

This functionality aims to resemble the behavior of the popular R package, afex, which features a `fun_aggregate` parameter. Quote from the [afex docs](https://cran.r-project.org/web/packages/afex/afex.pdf):

> fun_aggregate
> 
> The function for aggregating the data before running the ANOVA if
> there is more than one observation per individual and cell of the
> design. The default NULL issues a warning if aggregation is necessary
> and uses mean. Pass mean directly to avoid the warning.

### CODE REVIEW

**Comment by josef-pkt** [comment]
@yl565 @jona-sassenhagen   Do you have comments to this PR?

I'm not really familiar with the theoretical background here.
@hoechenberger Can you add unit tests against the R package to verify one or a few cases?

The main thing that I am a bit suspicious about is that using the mean of observation when not all cells have the same number of repeats would cause heteroscedasticity. The variance in this case would be the variance of the mean and differ across aggregated observations. In that case either the model or the inference would need to correct for heteroscedasticity, as far as I understand from the general cases.
(My guess is that it would violate something like the balancedness assumption in the current version.)

**Comment by hoechenberger** [comment]
@josef-pkt 

Collapsing the data, typically to means, is common practice in (RM-)Anova. I checked the [docs](https://cran.r-project.org/web/packages/ez/ez.pdf) of another R Anova package, `ez`, and it is stated there:

> Prior to running (though after obtaining running ANCOVA regressions as described in the details section), dv is collapsed to a mean for each cell defined by the combination of wid and any variables supplied to within and/or between and/or diff. Users are warned that while convenient when used properly, this automatic collapsing can lead to inconsistencies if the pre-collapsed data are unbal- anced (with respect to cells in the full design) and only the partial design is supplied to ezANOVA. When this is the case, use within_full to specify the full design to ensure proper automatic col- lapsing.

So you may have a point regarding heteroscedasticity.

However, it is quite common to have different numbers of observations per experimental condition and subjects in many psychological studies, because subjects sometimes fail to respond; or there is a problem with stimulus delivery; or data acquisition stops in the middle of an experiment, etc; yet the data is too valuable to simply throw it away.  So everything is somehow collapsed on a per-participant level before feeding it into the Anova. [Moscatelli et al. (2012)](http://jov.arvojournals.org/article.aspx?articleid=2192152) call this approach "parameter as outcome model" (PAOM).

I could certainly add unit tests against R's `ezANOVA` results, yes.

**Comment by hoechenberger** [comment]
> (My guess is that it would violate something like the balancedness assumption in the current version.)

The current implementation checks if the data is in fact balanced. This check is not removed; it is now simply performed _after_ the data have been aggregated.

**Comment by hoechenberger** [comment]
Mh, there are tests failing because `pandas.testing` cannot be imported. I don't know with which pandas version this module first appeared … It's definitely present in 0.20.1.

**Comment by hoechenberger** [comment]
@josef-pkt 

Quoting  Moscatelli et al. (2012):

>  In the case of repeated responses from more than one subject, ordinary GLMs treat the errors within subject in the same manner as the errors between subjects, and tend to produce invalid standard errors of the estimated parameters. A typical approach to overcome this problem consists in applying a two-level analysis (e.g., Morrone, Ross, & Burr, 2005; Pariyadath & Eagleman, 2007; Johnston et al., 2008). First, the parameters of the psychometric function are estimated for each subject. Next, the individual estimates are pooled to perform the second-level analysis, and inference is carried out, for example, by means of t test or ANOVA statistics.

**Comment by josef-pkt** [comment]
Thanks for the references
The enhancement is fine with me given that the R packages do the same, and if we have the warnings. Maybe the default should be to raise an exception to force users to explicitly request this.

(One main issue for the anova rm case is whether we can provide something with "cleaner" statistical properties in the unbalanced case that is also easy to use. However, I have not much background on the details here.)

**Comment by josef-pkt** [comment]
about the pandas test assert: 
we might have to make this conditional on the pandas version and use a simple numpy assert on the data for older pandas version, `assert_allclose(np.asarray(dataframe), ....)`

**Comment by hoechenberger** [comment]
Thanks @josef-pkt. Do you still want me to add tests to compare to R results?
How exactly would I do that, simply copy values from R and hard-code them into the test case? Or actually interact with an R session (using `rpy2`)?

I'm unfortunately not an expert on RM-Anova (or statistic in general) either; just trying to follow "common practice" in my field of science.

**Comment by hoechenberger** [comment]
> Maybe the default should be to raise an exception to force users to explicitly request this.

Good point, will change that.

**Comment by josef-pkt** [comment]
> How exactly would I do that, simply copy values from R and hard-code them into the test case?

Yes, that's what we do for most test cases, with verified values either from R, Stata, other packages or publications. (A problem is if the reference packages don't have the same functions but which is not the case here.)

**Comment by hoechenberger** [comment]
I'm a bit at a loss as to why Travis fails for some environments, but not for others…

**Comment by hoechenberger** [comment]
@josef-pkt 
I think this is good now. Want me to squash rebase & commits?

**Comment by josef-pkt** [comment]
I will go over it later today, then we decide on squashing

File: statsmodels/stats/anova.py [hunk 1]
```diff
@@ -432,18 +432,32 @@ class AnovaRM(object):
         The within-subject factors
     between : a list of string(s)
         The between-subject factors, this is not yet implemented
+    aggregate_func : None, 'mean', or function
+        If the data set contains more than a single observation per subject
+        and cell of the specified model, this function will be used to
+        aggregate the data, and a warning will be issued. 'mean' will
```
Comment (josef-pkt) [inline statsmodels/stats/anova.py]: Warning is now not issued

File: statsmodels/stats/anova.py [hunk 2]
```diff
@@ -455,8 +469,30 @@ def __init__(self, data, depvar, subject, within=None, between=None):
...
+                       'or pass the `aggregate_func` parameter.')
+                raise ValueError(msg)
+
         self._check_data_balanced()
 
+    def _aggregate(self):
+        self.data = (self.data
+                     .groupby([self.subject] + self.within,
+                              as_index=False)
+                     [self.depvar]
```
Comment (josef-pkt) [inline statsmodels/stats/anova.py]: Wouldn't this be better on the previous line

File: statsmodels/stats/tests/test_anova_rm.py [hunk 3]
```diff
@@ -131,3 +134,62 @@ def test_repeated_measures_collinearity():
...
+def test_repeated_measures_aggregate_func():
+    assert_raises(ValueError, AnovaRM, data.append(data), 'DV', 'id',
+                  within=['A', 'B', 'D'])
+
+    m1 = AnovaRM(data.append(data), 'DV', 'id', within=['A', 'B', 'D'],
+                 aggregate_func=np.mean)
+    m2 = AnovaRM(data.append(data), 'DV', 'id', within=['A', 'B', 'D'],
+                 aggregate_func=np.median)
+
+    assert m1.aggregate_func is not m2.aggregate_func
```
Comment (josef-pkt) [inline statsmodels/stats/tests/test_anova_rm.py]: Use numpy testing assert_, no plain assert in unit tests
Also afaics the results in m1 and m2 should be the same, additional test?

File: statsmodels/stats/tests/test_anova_rm.py [hunk 4]
```diff
@@ -131,3 +134,62 @@ def test_repeated_measures_collinearity():
...
+
+
+def test_repeated_measures_aggregate_func_mean():
+    m1 = AnovaRM(data.append(data), 'DV', 'id', within=['A', 'B', 'D'],
+                 aggregate_func=np.mean)
+
+    m2 = AnovaRM(data.append(data), 'DV', 'id', within=['A', 'B', 'D'],
+                 aggregate_func='mean')
+
+    assert m1.aggregate_func is m2.aggregate_func
```
Comment (josef-pkt) [inline statsmodels/stats/tests/test_anova_rm.py]: Same as above

**Comment by hoechenberger** [comment]
@josef-pkt I have addressed all your points.

**Comment by josef-pkt** [comment]
looks good to me now and looks ready for merge

github doesn't show the network or parent of the starting point of a PR anymore. (at least I don't find it anymore)
@hoechenberger  Is this (re)based on current master?

**Comment by hoechenberger** [comment]
> @hoechenberger Is this (re)based on current master?

Yes, this is rebased on master with all commits squashed into a single one :)

**Comment by josef-pkt** [comment]
@hoechenberger  Thank you, merging

**Comment by josef-pkt** [comment]
@hoechenberger This PR went very well. If you have more improvement, then please keep them coming.

**Comment by hoechenberger** [comment]
@josef-pkt Great, thank you! :)

**Comment by jona-sassenhagen** [comment]
Sorry for missing this! But everything @hoechenberger did agrees with my (limited) understanding of the matter.

**Pull request merged**